### PR TITLE
Runtime API: aggregate block messages | Introduce Starknet OS output

### PR DESCRIPTION
**Pull request by m-kus**
1. Messages that are sent to L1 or consumed from L1 during a particular block now can be aggregated via runtime API
2. `StarknetOsOutput` type was introduced with a binary codec compatible with Solidity contract (tested on mainnet data)

Related to #1133 

On the client we can get program output roughly as follows:

```rust
    fn get_snos_program_output(&self, block_id: BlockId) -> RpcResult<StarknetOsOutput> {
        let substrate_block_hash = self.substrate_block_hash_from_starknet_block(block_id).map_err(|e| {
            error!("'{e}'");
            StarknetRpcApiError::BlockNotFound
        })?;

        let block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash).unwrap_or_default();

        let old_root = if block.header().block_number > 0 {
            let parent_block_hash = (TryInto::<FieldElement>::try_into(block.header().parent_block_hash)).unwrap();
            let substrate_parent_block_hash =
                self.substrate_block_hash_from_starknet_block(BlockId::Hash(parent_block_hash)).map_err(|e| {
                    error!("'{e}'");
                    StarknetRpcApiError::BlockNotFound
                })?;

            let parent_block =
                get_block_by_block_hash(self.client.as_ref(), substrate_parent_block_hash).unwrap_or_default();
            parent_block.header().global_state_root.into()
        } else {
            FieldElement::default()
        };

        let block_extrinsics = self
            .client
            .block_body(substrate_block_hash)
            .map_err(|e| {
                error!("'{e}'");
                StarknetRpcApiError::InternalServerError
            })?
            .ok_or(StarknetRpcApiError::BlockNotFound)?;
        let chain_id = self.chain_id()?.0.into();

        let (messages_to_l1, messages_to_l2) = self
            .client
            .runtime_api()
            .get_starknet_messages(substrate_block_hash, block_extrinsics, chain_id)
            .map_err(|e| {
                error!("'{e}'");
                StarknetRpcApiError::InternalServerError
            })?;

        Ok(StarknetOsOutput {
            prev_state_root: old_root,
            new_state_root: block.header().global_state_root.into(),
            block_number: block.header().block_number,
            block_hash: block.header().hash::<H>().into(),
            config_hash: StarkFelt::ZERO,
            messages_to_l1,
            messages_to_l2,
        })
    }
```

### CODE REVIEW

File: crates/runtime/src/lib.rs [hunk 1]
```diff
@@ -356,8 +357,45 @@ impl_runtime_apis! {
...
+                        Some(transaction.compute_hash::<<Runtime as Config>::SystemHash>(chain_id, false))
+                    },
+                    RuntimeCall::Starknet( consume_l1_message { transaction, .. }) => {
+                        let tx_hash = Some(transaction.compute_hash::<<Runtime as Config>::SystemHash>(chain_id, false));
+                        messages_to_l2.push(transaction.into());
+                        tx_hash
+                    },
+                    _ => None
+                };
+                if let Some(tx_hash) = tx_hash {
```
Comment (EvolveArt) [inline crates/runtime/src/lib.rs:390]: I don't think declare/deploy transactions can have messages tho

File: starknet-rpc-test/src/constants.rs [hunk 2]
```diff
@@ -25,7 +26,7 @@ pub const TOKEN_CLASS_HASH: &str = "0x000000000000000000000000000000000000000000
 pub const ARGENT_CONTRACT_ADDRESS: &str = "0x0000000000000000000000000000000000000000000000000000000000000002";
 
 // Taken from https://github.com/0xSpaceShard/starknet-devnet-rs/blob/main/crates/starknet-server/tests/common/mod.rs#L5
-pub const MIN_PORT: u16 = 1025;
+pub const MIN_PORT: u16 = 49_152;
```
Comment (EvolveArt) [inline starknet-rpc-test/src/constants.rs:29]: please update the above comment

File: starknet-rpc-test/src/constants.rs [hunk 3]
```diff
@@ -7,6 +7,7 @@ pub const CAIRO_1_ACCOUNT_CONTRACT_CLASS_HASH: &str =
     "0x35ccefcf9d5656da623468e27e682271cd327af196785df99e7fee1436b6276";
 pub const ERC20_CAIRO_0_CONTRACT: &str = "0x040e59c2c182a58fb0a74349bfa4769cbbcba32547591dd3fb1def8623997d00";
 pub const SEQUENCER_ADDRESS: &str = "0xdead";
+pub const UDC_ADDRESS: &str = "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf";
```
Comment (EvolveArt) [inline starknet-rpc-test/src/constants.rs:10]: need to update the readme + it's linked to this #1100 I think

File: starknet-rpc-test/get_transaction_receipt.rs [hunk 4]
```diff
@@ -265,3 +268,91 @@ async fn fail_invalid_transaction_hash(#[future] madara: MadaraClient) -> Result
...
+            assert_eq_msg_to_l1(
+                vec![MsgToL1 {
+                    from_address: contract_address,
+                    to_address: FieldElement::ONE, // hardcoded in the contract
+                    payload: vec![],               // no payload
+                }],
+                receipt.messages_sent,
+            );
+        }
+        _ => panic!("expected tx receipt"),
```
Comment (EvolveArt) [inline starknet-rpc-test/get_transaction_receipt.rs:354]: can we use assert_matches here?

File: starknet-rpc-test/get_transaction_receipt.rs [hunk 5]
```diff
@@ -265,3 +268,91 @@ async fn fail_invalid_transaction_hash(#[future] madara: MadaraClient) -> Result
...
+
+    let invoke_tx = account.call_contract(contract_address, "send_message_l2_to_l1", vec![], None);
+
+    txs = madara.create_block_with_txs(vec![Transaction::Execution(invoke_tx)]).await?;
+
+    // 6. Finally, checking that there is a single MessageToL1 in the receipt
+
+    let invoke_tx_hash = match txs.remove(0).unwrap() {
+        TransactionResult::Execution(rpc_response) => rpc_response.transaction_hash,
+        _ => panic!("expected execution result"),
```
Comment (EvolveArt) [inline starknet-rpc-test/get_transaction_receipt.rs:338]: assert_matches

File: starknet-rpc-test/get_transaction_receipt.rs [hunk 6]
```diff
@@ -265,3 +268,91 @@ async fn fail_invalid_transaction_hash(#[future] madara: MadaraClient) -> Result
 
     Ok(())
 }
+
+#[rstest]
+#[tokio::test]
+async fn work_with_messages_to_l1(#[future] madara: MadaraClient) -> Result<(), anyhow::Error> {
```
Comment (EvolveArt) [inline starknet-rpc-test/get_transaction_receipt.rs:274]: just need to add tests for edge cases:
- What happens if the tx fail ?
- Is there any limit on the number of messages per tx ?

**Review by EvolveArt — CHANGES_REQUESTED** [review]
left few comments

File: crates/runtime/src/lib.rs (hunk 1 above)
Comment (tdelabro) [reply crates/runtime/src/lib.rs:390]: Doesn't it depend on the account contract used? 
Most probably don't, but is it really impossible?

File: crates/pallets/starknet/src/runtime_api.rs [hunk 7]
```diff
@@ -62,6 +63,8 @@ sp_api::decl_runtime_apis! {
         fn get_tx_execution_outcome(tx_hash: TransactionHash) -> Option<Vec<u8>>;
         /// Return messages sent to L1 during tx execution
         fn get_tx_messages_to_l1(tx_hash: TransactionHash) -> Vec<MessageToL1>;
+        /// Return two lists of messages sent (to L1) and consumed (from L1) during this block
+        fn get_starknet_messages(block_extrinsics: Vec<<Block as BlockT>::Extrinsic>, chain_id: Felt252Wrapper) -> (Vec<MessageL2ToL1>, Vec<MessageL1ToL2>);
```
Comment (tdelabro) [inline crates/pallets/starknet/src/runtime_api.rs:67]: What dictates the signature return type? What will it be used by

File: crates/runtime/src/lib.rs (hunk 1 above)
Comment (m-kus) [reply crates/runtime/src/lib.rs:390]: I though abt it, technically there should be no limitations? Same goes to the fee execution

Comment (tdelabro) [reply crates/runtime/src/lib.rs:390]: The fee execution is done through a hardcoded contract configured by the appchain maintainers. So we can know for sure it does not produce a message.

File: crates/pallets/starknet/src/runtime_api.rs (hunk 7 above)
Comment (m-kus) [reply crates/pallets/starknet/src/runtime_api.rs:67]: Starknet os output struct has two fields with the respectful types, I added a code snippet to illustrate that (pr description)

**Comment by tdelabro** [comment]
@m-kus I merged the other one, you should rebase more probably cherry-pick your commits and push force on this one :)

File: starknet-rpc-test/src/constants.rs (hunk 3 above)
Comment (m-kus) [reply starknet-rpc-test/src/constants.rs]: Which readme file?

File: crates/primitives/snos-output/src/codec.rs [hunk 8]
```diff
@@ -0,0 +1,181 @@
+use alloc::vec::Vec;
+
+use parity_scale_codec::{Decode, Encode, Input, Output};
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/codec.rs:3]: I have the feeling you are using the `parity_scale_codec` `Encode` and `Decode` traits, but the way you implement them is not using the [SCALE encoding](https://docs.substrate.io/reference/scale-codec/).

Am I right? If so, you should not use those traits, and instead create some trait or method of your own for this purpose.

File: crates/primitives/snos-output/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+#[cfg(test)]
+mod tests;
+
+use alloc::vec::Vec;
+
+use starknet_api::api_core::{ContractAddress, EntryPointSelector, EthAddress, Nonce};
+use starknet_api::hash::{StarkFelt, StarkHash};
+
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+#[cfg_attr(feature = "scale-info", derive(scale_info::TypeInfo))]
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/lib.rs:21]: Add doc

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/primitives/snos-output/src/codec.rs (hunk 8 above)
Comment (m-kus) [reply crates/primitives/snos-output/src/codec.rs:3]: Yep, that was just the easiest way to both use types in runtime api and abi encode them.
I will derive scale trait then and add a custom serialization trait implementation

File: starknet-rpc-test/get_transaction_receipt.rs (hunk 6 above)
Comment (tdelabro) [reply starknet-rpc-test/get_transaction_receipt.rs]: > What happens if the tx fail ?

I don't get what you mean. If the tx fail there is not tx receipt

File: crates/primitives/snos-output/src/codec.rs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    }
+}
+
+fn segment_append_to<T: SnosCodec>(segment: Vec<T>, output: &mut Vec<StarkFelt>) {
+    let segment_size = segment.iter().map(|msg| msg.felt_len() as u64).sum::<u64>();
+    output.push(segment_size.into());
+    for item in segment {
+        item.append_to(output);
+    }
+}
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/codec.rs:33]: You should 
```rust
impl SnosCodec for Vec<T>
  where T: SnosCodec
{
  ...
}
```  

And the reuse it instead of 
```
segment_append_to(self.messages_to_l1, output);
segment_append_to(self.messages_to_l2, output);
```
and

```
output.push((self.payload.len() as u64).into());
let mut payload = self.payload;
output.append(&mut payload);
```

File: crates/primitives/snos-output/src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+use crate::codec::SnosCodec;
+use crate::conversions::eth_address_to_felt;
+use crate::{MessageL1ToL2, MessageL2ToL1, StarknetOsOutput};
+
+trait Decode {
+    fn decode<I: Read>(input: &mut I) -> Self;
+}
+
+fn segment_decode<T: Decode + SnosCodec, I: Read>(input: &mut I) -> Vec<T> {
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/tests.rs:14]: `Impl Decode for Vec<T: Decode> {...}`

Comment (tdelabro) [reply crates/primitives/snos-output/src/tests.rs:14]: `impl Decode for Vec<StarkFelt>` would come for free

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/primitives/snos-output/src/codec.rs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+///     2. Get onchain data hash & size
+///     3. ABI encode parameters 
+///  
+/// NOTE: Field element (252 bit) is encoded as an EVM word (256 bit) and vice versa
+/// EVM developer should be aware of that and prevent data loss by not using the higest 4 bits
+pub trait SnosCodec: Sized {
+    fn felt_len(&self) -> usize;
+    fn append_to(self, output: &mut Vec<StarkFelt>);
+
+    fn as_vec(self) -> Vec<StarkFelt> {
```
[the remark this answers is not in the document]
Comment (m-kus) [reply crates/primitives/snos-output/src/codec.rs:20]: Ok :)

File: crates/primitives/snos-output/src/codec.rs (hunk 10 above)
Comment (m-kus) [reply crates/primitives/snos-output/src/codec.rs:33]: It's a bit different for segment and Vec<StarkFelt>, in the fist case we deal with dynamically sized elements

File: crates/primitives/snos-output/src/codec.rs [hunk 13]
```diff
@@ -5,76 +5,88 @@ use starknet_api::hash::StarkFelt;
...
 ///     2. Get onchain data hash & size
-///     3. ABI encode parameters 
+///     3. ABI encode parameters
 ///  
 /// NOTE: Field element (252 bit) is encoded as an EVM word (256 bit) and vice versa
 /// EVM developer should be aware of that and prevent data loss by not using the higest 4 bits
 pub trait SnosCodec: Sized {
-    fn felt_len(&self) -> usize;
-    fn append_to(self, output: &mut Vec<StarkFelt>);
+    fn size_hint(&self) -> usize;
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/codec.rs:17]: Add doc for each method

File: crates/primitives/snos-output/src/tests.rs [hunk 14]
```diff
@@ -16,7 +16,7 @@ fn segment_decode<T: Decode + SnosCodec, I: Read>(input: &mut I) -> Vec<T> {
     let mut items: Vec<T> = Vec::new();
     while segment_len > 0 {
         let item = T::decode(input);
-        segment_len -= item.felt_len();
+        segment_len -= item.size_hint();
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/tests.rs:19]: I think you should be using the actual item size instead of `size_hint`.
For parity_scale_codec, `size_hint` is just a hint here to optimize memory allocation, not something you should rely on 100%. It can be an erroneous value.

File: crates/primitives/snos-output/src/tests.rs [hunk 15]
```diff
@@ -16,7 +16,7 @@ fn segment_decode<T: Decode + SnosCodec, I: Read>(input: &mut I) -> Vec<T> {
     let mut items: Vec<T> = Vec::new();
     while segment_len > 0 {
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/tests.rs:17]: "vec" instead of "segment"

File: crates/primitives/snos-output/src/tests.rs (hunk 14 above)
Comment (tdelabro) [reply crates/primitives/snos-output/src/tests.rs:19]: Or you can change this for your trait. But you should rename the method to not be a `hint` anymore and explain in the doc how the user should impl it

File: crates/primitives/snos-output/src/codec.rs [hunk 16]
```diff
@@ -5,76 +5,88 @@ use starknet_api::hash::StarkFelt;
...
+    fn encode_to(self, output: &mut Vec<StarkFelt>);
 
-    fn as_vec(self) -> Vec<StarkFelt> {
-        let mut output: Vec<StarkFelt> = Vec::with_capacity(self.felt_len());
-        self.append_to(&mut output);
+    fn into_vec(self) -> Vec<StarkFelt> {
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/codec.rs:20]: `into_encodec_vec` to be more explicit

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/primitives/snos-output/src/codec.rs (hunk 16 above)
Comment (tdelabro) [reply crates/primitives/snos-output/src/codec.rs:20]: otherwise, we lose the dimension of data transformation

File: crates/primitives/snos-output/src/codec.rs [hunk 17]
```diff
@@ -5,76 +5,88 @@ use starknet_api::hash::StarkFelt;
...
-        output.push(self.prev_state_root);
-        output.push(self.new_state_root);
-        output.push(self.block_number);
-        output.push(self.block_hash);
-        output.push(self.config_hash);
-        segment_append_to(self.messages_to_l1, output);
-        segment_append_to(self.messages_to_l2, output);
+    fn encode_to(self, output: &mut Vec<StarkFelt>) {
+        let segment_size = self.size_hint() as u64;
+        output.push(segment_size.into());
```
[the remark this answers is not in the document]
Comment (m-kus) [reply crates/primitives/snos-output/src/codec.rs:44]: Yep, that's how segment is encoded

File: crates/primitives/snos-output/src/tests.rs (hunk 14 above)
Comment (m-kus) [reply crates/primitives/snos-output/src/tests.rs:19]: That's why I named it differently :) ok

File: crates/primitives/snos-output/src/codec.rs [hunk 18]
```diff
@@ -5,76 +5,88 @@ use starknet_api::hash::StarkFelt;
...
     }
 }
 
-impl SnosCodec for StarknetOsOutput {
-    fn felt_len(&self) -> usize {
-        7 + self.messages_to_l1.iter().map(|msg| msg.felt_len()).sum::<usize>()
-            + self.messages_to_l2.iter().map(|msg| msg.felt_len()).sum::<usize>()
+impl<T: SnosCodec> SnosCodec for Vec<T> {
+    fn size_hint(&self) -> usize {
+        self.iter().map(|elt| elt.size_hint()).sum()
```
[the remark this answers is not in the document]
Comment (m-kus) [reply crates/primitives/snos-output/src/codec.rs]: nope, segment size does not include the one used for encoding it

File: crates/primitives/snos-output/src/felt_reader.rs [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+            return Err(FeltReaderError::OutOfBounds);
+        }
+
+        let res = self.data[self.offset];
+        self.offset += 1;
+
+        Ok(res)
+    }
+
+    pub fn read_segment(&mut self) -> Result<FeltReader, FeltReaderError> {
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/felt_reader.rs:55]: This method is not actually reading the segment.
It initializes another Reader for a subset of itself. 
The name is misleading.

File: crates/primitives/snos-output/src/codec.rs [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+        output.push(self.selector);
+        self.payload.encode_to(output);
+    }
+
+    fn decode(input: &mut FeltReader) -> Result<Self, FeltReaderError> {
+        Ok(Self {
+            from_address: input.read()?,
+            to_address: input.read()?,
+            nonce: input.read()?,
+            selector: input.read()?,
```
Comment (tdelabro) [inline crates/primitives/snos-output/src/codec.rs:103]: You should be calling `.decode()` for each one of those lines.
Here you are manually doing the inline of the impl of decode for StarkFelt.
But you are not supposed to know about the internal impl

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/primitives/snos-output/src/felt_reader.rs (hunk 19 above)
Comment (tdelabro) [reply crates/primitives/snos-output/src/felt_reader.rs:55]: Also, `read` should probably be `read_one`.

**Review by tdelabro — APPROVED** [review]

**Review by EvolveArt — APPROVED** [review]
